akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterShardingEntityProvider.scala

```diff
@@ -0,0 +1,141 @@
+      case messageExtractor                                             =>
+        // fall back to number of shards defined in settings
+        val numberOfShards = ClusterShardingSettings(system).numberOfShards
+        new ShardingMessageExtractor[E, M] {
+          override def entityId(message: E): String = messageExtractor.entityId(message)
+          override def shardId(entityId: String): String = HashCodeMessageExtractor.shardId(entityId, numberOfShards)
+          override def unwrapMessage(message: E): M = messageExtractor.unwrapMessage(message)
+        }
+    }
+
+    val shardedEntity: ShardedEntity[M, E] =
+      new ShardedEntity(
+        ctx => entity.createBehavior(toEntityContext(ctx)),
+        toShardedEntityTypeKey(entity.typeKey),
+        entity.stopMessage,
+        entity.entityProps,
+        clusterSettings,
+        shardingMessageExtractor,
+        None,
+        None,
+        None)
+
+    sharding.init(shardedEntity)
+  }
+
+  override def entityRefFor[M](typeKey: EntityTypeKey[M], entityId: String): EntityRef[M] =
+    toEntityRef(sharding.entityRefFor(toShardedEntityTypeKey(typeKey), entityId))
+
+}
+
+private object EntityAdapter {
+
+  def toShardedEntityTypeKey[M](typeKey: EntityTypeKey[M]): ShardedEntityTypeKey[M] =
+    typeKey match {
+      case shardedTypeKey: ShardedEntityTypeKey[M] @unchecked => shardedTypeKey
+      case typeKeyImpl: akka.actor.typed.EntityTypeKeyImpl[M] @unchecked =>
+        ShardedEntityTypeKeyImpl(typeKey.name, typeKeyImpl.messageClassName)
+    }
+
+  def toEntityContext[M](ctx: ShardedEntityContext[M]): EntityContext[M] = {
+    val typeKey = toEntityTypeKey(ctx.entityTypeKey)
+    // FIXME: needs proper conversion
+    val shardRef: ActorRef[EntityCommand] = ctx.shard.asInstanceOf[ActorRef[EntityCommand]]
+    new EntityContext(typeKey, ctx.entityId, shardRef)
+  }
+
+  def toEntityTypeKey[M](typeKey: ShardedEntityTypeKey[M]): EntityTypeKey[M] = {
+    val impl = typeKey.asInstanceOf[ShardedEntityTypeKeyImpl[M]]
+    akka.actor.typed.EntityTypeKeyImpl[M](typeKey.name, impl.messageClassName)
+  }
+
+  def toEntityRef[M](shardedEntity: ShardedEntityRef[M]): EntityRef[M] = new ShardedEntityRefWrapper(shardedEntity)
+
+  private[akka] class ShardedEntityRefWrapper[-M](delegate: ShardedEntityRef[M])
+      extends EntityRef[M]
+      with InternalRecipientRef[M] {
+
+    override def entityId: String = delegate.entityId
+
+    override def typeKey: EntityTypeKey[M] = toEntityTypeKey(delegate.typeKey)
+
+    override def tell(msg: M): Unit =
+      msg match {
+        case EntityEnvelope.StartEntity(entityId) =>
+          delegate.tell(ClassicStartEntity(entityId).asInstanceOf[M])
+        case entityEnv: EntityEnvelope[_] =>
+          val shardEnv = new ShardingEnvelope(entityEnv.entityId, entityEnv.message)
+          delegate.tell(shardEnv.asInstanceOf[M])
+        case _ => delegate.tell(msg)
+      }
```

Comment:
Actually, this is what will allow us to remove the serialization. We can just convert to ShardingEnvelope here.